Improve performance for 17 rules

andrei-epure-sonarsource opened the pull request:
Related to #3768 (another small change will be made tomorrow)
Fixes #3767
Fixes #3766 !!
Fixes #3764
Fixes #3762
Fixes #3760
Partial fix for #3731 (another small change tomorrow)

andrei-epure-sonarsource commented:
@costin-zaharia-sonarsource I still have to fix the code smells, bringing back to in progress

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs:
```diff
@@ -95,20 +95,16 @@ private static void CheckCountCall(SyntaxNodeAnalysisContext context)
-            if (!(context.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol) ||
-                methodSymbol.Name != CountName ||
-                invocation.ArgumentList == null ||
+            if (invocation.ArgumentList == null ||
```
I would try to simplify this a bit. Maybe something similar with:
```
var invocation = (InvocationExpressionSyntax)context.Node;
if (invocation.ArgumentList?.Arguments.Count == 0 &&
    invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    context.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol &&
    methodSymbol.Name == CountName &&
    methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T) &&
    context.SemanticModel.GetTypeInfo(memberAccess.Expression).Type.GetMembers(CountName).OfType<IPropertySymbol>().Any())
{
    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, GetReportLocation(invocation),
                                                         string.Format(MessageUseInstead, $"'{CountName}' property")));
}
```

and maybe extract `context.SemanticModel.GetTypeInfo(memberAccess.Expression).Type.GetMembers(CountName).OfType<IPropertySymbol>().Any()` to a method (can be local)

In this way we have a single list with preconditions and avoid all the negated checks. It is also more concise.

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EnumerableSumInUnchecked.cs:
```diff
@@ -46,21 +46,20 @@ protected override void Initialize(SonarAnalysisContext context)
                     }
 
-                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
+                    if (IsSumInsideUnchecked(invocation))
```
Similar with previous cases, the expression body can be changed with:
```
var invocation = (InvocationExpressionSyntax)c.Node;
if (invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    IsSumInsideUnchecked(invocation))
{
    var methodSymbol = c.SemanticModel.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
    if (IsSumOnInteger(methodSymbol))
    {
        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
    }
}
```

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetTypeWithIsAssignableFrom.cs:
```diff
@@ -92,12 +97,16 @@ protected override void Initialize(SonarAnalysisContext context)
-            var typeExpression = context.SemanticModel.GetTypeInfo(binary.Left).Type;
-            var typeCastTo = context.SemanticModel.GetTypeInfo(binary.Right).Type;
 
+            var typeExpression = context.SemanticModel.GetTypeInfo(binary.Left).Type;
```
What do you think about using the following snippet instead?
```
if (context.SemanticModel.GetTypeInfo(binary.Left).Type is {} leftType &&
    leftType.IsClass() &&
    context.SemanticModel.GetTypeInfo(binary.Right).Type is {} rightType &&
    rightType.IsClass() &&
    !rightType.Is(KnownType.System_Object) &&
    leftType.DerivesOrImplements(rightType))
{
    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, binary.GetLocation(), MessageNullCheck));
}
```

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetTypeWithIsAssignableFrom.cs:
```diff
@@ -129,13 +138,14 @@ private static void CheckIfIsExpressionCanBeReplacedByNullCheck(SyntaxNodeAnalys
 
-            var typeSyntax = (sideB as TypeOfExpressionSyntax)?.Type;
-            if (typeSyntax == null)
+            if (!sideA.ToString().Contains("GetType") ||
```
I think we can move this up since checking for string is cheap and we can do this before retrieving the `sideB` type.

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/OrderByRepeated.cs:
```diff
@@ -20,5 +20,27 @@ public void Test()
+        }
+
+        public void Foo() { }
+
```
pedantic: empty line

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LiteralSuffixUpperCase.cs:
```diff
@@ -28,7 +28,7 @@ public void Test(long ui)
         {
             const uint a = 0U;
-            const long b = 0L;
+            const long b = 100L;
```
was this change needed?

andrei-epure-sonarsource replied:
it wasn't needed, however the test case has only 1-digit integer numbers, so I wanted to add a multi-digit test case

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InvocationResolvesToOverrideWithParams.cs:
```diff
@@ -62,10 +62,11 @@ protected override void Initialize(SonarAnalysisContext context)
                 !invokedMethodSymbol.Parameters.Any() ||
-                !invokedMethodSymbol.Parameters.Last().IsParams ||
-                argumentList == null)
+                !invokedMethodSymbol.Parameters.Last().IsParams)
```
We can include in this "if" the condition from next one to remove one extra exit point from this method.

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:
```diff
@@ -48,14 +49,17 @@ protected override void Initialize(SonarAnalysisContext context)
 
-                    if (text[text.Length - 1] == 'l' &&
-                        c.SemanticModel.GetTypeInfo(literal).Type.Is(KnownType.System_Int64))
+                    if (text[text.Length - 1] == 'l' && NotUnsigned(text))
```
We can use here [EndsWith(Char)
](https://docs.microsoft.com/en-us/dotnet/api/system.string.endswith?view=net-5.0#System_String_EndsWith_System_Char_)

andrei-epure-sonarsource replied:
it was added in .NET Standard 2.1, and we're still on 2.0.

costin-zaharia-sonarsource replied:
Then maybe the [overload with string](https://docs.microsoft.com/en-us/dotnet/api/system.string.endswith?view=netframework-4.8#overloads)? I think it's easier to read.

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:
```diff
@@ -48,14 +49,17 @@ protected override void Initialize(SonarAnalysisContext context)
         }
+
+        // a number may be `2ul` or `2Ul` and it's easier to read
+        private static bool NotUnsigned(string text) =>
```
Nice! Since we subscribe to `NumericLiteralExpression` there is no need for `SemanticModel` calls to get the type :) 

Can we use `text.Length >= 3`. I think that if it's a number it can't start with a char so the condition can't be true for length = 2.

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/OrderByRepeated.cs:
```diff
@@ -46,17 +46,22 @@ protected override void Initialize(SonarAnalysisContext context)
                 {
                     var outerInvocation = (InvocationExpressionSyntax)c.Node;
-                    if (!IsMethodOrderByExtension(outerInvocation, c.SemanticModel))
+                    if (!(outerInvocation.Expression is MemberAccessExpressionSyntax memberAccess))
```
We can simplify the lambda body to:
```
var outerInvocation = (InvocationExpressionSyntax)c.Node;
if (outerInvocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    memberAccess.Expression is InvocationExpressionSyntax innerInvocation &&
    IsMethodOrderByExtension(outerInvocation, c.SemanticModel) &&
    (IsMethodOrderByExtension(innerInvocation, c.SemanticModel) || IsMethodThenByExtension(innerInvocation, c.SemanticModel)))
{
    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
}
```

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs:
```diff
@@ -47,21 +47,22 @@ protected override void Initialize(SonarAnalysisContext context)
-                        methodSymbol.Name != "ToCharArray" ||
-                        !methodSymbol.IsInType(KnownType.System_String) ||
-                        methodSymbol.Parameters.Length != 0)
+                    if (!(invocation.Parent is ElementAccessExpressionSyntax) &&
```
Can be simplified to something similar with:
```
var invocation = (InvocationExpressionSyntax)c.Node;
if ((invocation.Parent is ElementAccessExpressionSyntax || invocation.Parent is ForEachStatementSyntax) &&
    invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    invocation.ToString().Contains("ToCharArray") &&
    c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol &&
    methodSymbol.Name == "ToCharArray" && methodSymbol.IsInType(KnownType.System_String) &&
    methodSymbol.Parameters.Length == 0)
{
    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
}
```

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs:
```diff
@@ -47,21 +47,22 @@ protected override void Initialize(SonarAnalysisContext context)
                     }
 
-                    if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+                    if (!invocation.ToString().Contains("ToCharArray") ||
```
This pattern appears in a lot of places, should we try to extract it or a part in order to enable reuse?

Also related to this, we have `KnownMethods` class which has some methods similar with this. Did you consider extracting the name check as a method there?

andrei-epure-sonarsource replied:
`KnownMethods` operates on symbols, but I extracted it in extension methods.

I think anyway `ToString` has its cost as well, it probably iterates over the AST and calls the sub-nodes `ToString`s ... but at least if we want to change the method and improve it, we'll have a central place to do it

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TypeExaminationOnSystemType.cs:
```diff
@@ -50,13 +50,18 @@ protected override void Initialize(SonarAnalysisContext context)
+                    if (invocationString.Contains("IsInstanceOfType") ||
+                        invocationString.Contains("GetType"))
                     {
-                        return;
-                    }
+                        if (!(c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol))
+                        {
+                            return;
+                        }
 
-                    CheckGetTypeCallOnType(invocation, methodSymbol, c);
-                    CheckIsInstanceOfTypeCallWithTypeArgument(invocation, methodSymbol, c);
+                        CheckGetTypeCallOnType(invocation, methodSymbol, c);
+                        CheckIsInstanceOfTypeCallWithTypeArgument(invocation, methodSymbol, c);
+                    }
```
We can reduce this to:
```
var invocation = (InvocationExpressionSyntax)c.Node;
var invocationString = invocation.ToString();

if ((invocationString.Contains("IsInstanceOfType") || invocationString.Contains("GetType")) &&
    c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol)
{
    CheckGetTypeCallOnType(invocation, methodSymbol, c);
    CheckIsInstanceOfTypeCallWithTypeArgument(invocation, methodSymbol, c);
}
```

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs:
```diff
@@ -57,27 +57,27 @@ protected HardcodedIpAddressBase(IAnalyzerConfiguration analyzerConfiguration) :
-                var variableName = GetAssignedVariableName(stringLiteral);
-                if (variableName != null &&
-                    IgnoredVariableNames.Any(variableName.Contains))
+                if (literalValue == "::" ||
```
I think this is less CPU intensive than `GetAssignedVariableName` and `HasAttributes` which have to visit the syntax tree to search for identifiers and attributes.

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs:
```diff
@@ -103,22 +103,35 @@ protected override void Initialize(SonarAnalysisContext context)
-                    }
                     var leftNode = GetLeftNode(addExpression);
+
+                    var isInCheckedContext = false;
+                    // SemanticModel calls are expensive
+                    var ranSemanticCheck = false;
```
We don't need 2 variables for this. This can be nicely done with `Lazy<bool>` and the code will simplify a lot.
```
var isInCheckedContext = new Lazy<bool>(() => IsInCheckedContext(addExpression, c.SemanticModel));
```

costin-zaharia-sonarsource reviewed (changes requested):
I think the changes are very nice and I can't wait for the measurement results!

I've suggested though some changes to simplify the code.

andrei-epure-sonarsource commented:
@costin-zaharia-sonarsource I addressed the comments.
Please don't merge yet, as based on the perf results we may need to adjust some of the rules changes.

costin-zaharia-sonarsource commented:
> Please don't merge yet, as based on the perf results we may need to adjust some of the rules changes.

I'll wait then to review the final changes.

andrei-epure-sonarsource commented:
I looked just at efcore for saving time. I did two runs for 8.14 and two runs for this version.
- AssertionArgsShouldBePassedInCorrectOrder went from 1902/1941 to 1574/1619
- BeginInvokePairedWithEndInvoke went from 33647/34181 to 3140/3159 (nodatime from 1388 to 250)
- CollectionQuerySimplification went from 13019/13032 to 11710/11815
- DebugAssertHasNoSideEffects went from 2145/2175 to 1926/1964 (for n2cms from 364 to 294)
- EnumerableSumInUnchecked went from 2140/2154 to 1554/1516 (for n2cms from 359 to 125)
- !!!!! FieldsShouldBeEncapsulatedInProperties from <0.001 to 107
- GetTypeWithIsAssignableFrom from 3905/3968 to 2750/2591 (for n2cms from 591 to 310)
- InvocationResolvesToOverrideWithParams from 4802 to 4165
- ! LiteralSuffixUpperCase from 237/247 to 236/243 (but it must depend on the number of NumericLiterals in the codebase!)
- OrderByRepeated from 5121 to 1989 (n2cms from 553 to 175, nodatime 613 to 152)
- ! RedundantCast from 1835/1900 to 1871/1870
- RedundantToCharArrayCall from from 3864 to 1611
- ReturnEmptyCollectionInsteadOfNull from 1837 to 901
- TypeExaminationOnSystemType from 3921 to 3186
- UseStringIsNullOrEmpty from 7699 to 5685
- HardcodedIpAddress from 2 to 2
- UriShouldNotBeHardcoded from 350 to 297

andrei-epure-sonarsource commented:
Total improvement for these rules as tested on the _efcore_ project - from 86,5 seconds to 42,6 seconds

costin-zaharia-sonarsource commented on sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:
```diff
@@ -48,14 +49,18 @@ protected override void Initialize(SonarAnalysisContext context)
 
-                    if (text[text.Length - 1] == 'l' &&
-                        c.SemanticModel.GetTypeInfo(literal).Type.Is(KnownType.System_Int64))
+                    if (text[text.Length - 1] == 'l' && !ShouldIgnore(text))
```
don't want to use `text.EndsWith("l")` ?

andrei-epure-sonarsource replied:
as discussed offline, it's readable as is

costin-zaharia-sonarsource reviewed (approved):
Nice! 🚀 

Just 2 small comments.

andrei-epure-sonarsource commented:
I confirm there was no regression in terms of raised issues on the 3 projects I tested on.
